File: src/main/java/com/DavidM1A2/afraidofthedark/common/spell/component/deliveryMethod/base/DeliveryTransitionStateBuilder.java
package com.DavidM1A2.afraidofthedark.common.spell.component.deliveryMethod.base;

import com.DavidM1A2.afraidofthedark.common.spell.Spell;
import net.minecraft.entity.Entity;
import net.minecraft.util.math.Vec3d;
import net.minecraft.world.World;
import org.apache.commons.lang3.Validate;

/**
 * Builder class used to build a DeliveryTransitionState object
 */
public class DeliveryTransitionStateBuilder
{
    private Spell spell = null;
    private int stageIndex = 0;
    private World world = null;
    private Vec3d position = null;
    private Vec3d direction = null;
    private Entity entity = null;

    /**
     * Sets the spell to transition for
     *
     * @param spell The spell to transition for
     * @return The builder
     */
    public DeliveryTransitionStateBuilder withSpell(Spell spell)
    {
        this.spell = spell;
        return this;
    }

    /**
     * Sets the spell index we're transitioning from
     *
     * @param stageIndex The spell index we're transitioning from
     * @return The builder
     */
    public DeliveryTransitionStateBuilder withStageIndex(int stageIndex)
    {
        this.stageIndex = stageIndex;
        return this;
    }

    /**
     * Sets the world the transition is happening in
     *
     * @param world The world the transition is happening in
     * @return The builder
     */
    public DeliveryTransitionStateBuilder withWorld(World world)
    {
        this.world = world;
        return this;
    }

    /**
     * Sets the position the transition is happening at
     *
     * @param position The position the transition is happening at
     * @return The builder
     */
    public DeliveryTransitionStateBuilder withPosition(Vec3d position)
    {
        this.position = position;
        return this;
    }

    /**
     * Sets the direction the transition is happening towards
     *
     * @param direction The direction the transition is happening towards
     * @return The builder
     */
    public DeliveryTransitionStateBuilder withDirection(Vec3d direction)
    {
        this.direction = direction;
        return this;
    }

    /**
     * Sets the world, position, direction, and entity based on the entity's position
     *
     * @param entity The entity to grab data from
     * @return The builder
     */
    public DeliveryTransitionStateBuilder withEntity(Entity entity)
    {
        this.entity = entity;
        this.world = entity.getEntityWorld();
        this.position = entity.getPositionVector();
        this.direction = entity.getLookVec();
        return this;
    }

    /**
     * Sets the state of this builder based on an existing state
     *
     * @param state The state to copy from
     * @return The builder
     */
    public DeliveryTransitionStateBuilder copyOf(DeliveryTransitionState state)
    {
        this.spell = state.getSpell();
        this.stageIndex = state.getStageIndex();
        this.world = state.getWorld();
        this.position = state.getPosition();
        this.direction = state.getDirection();
        this.entity = state.getEntity();
        return this;
    }

    /**
     * Finishes the state object and returns it
     *
     * @return The state object
     */
    public DeliveryTransitionState build()
    {
        Validate.notNull(spell, "Spell must not be null!");
        Validate.notNull(world, "World must not be null!");
        Validate.notNull(position, "Position must not be null!");

        return new DeliveryTransitionState(spell, stageIndex, world, position, direction, entity);
    }
}

File: src/main/java/com/davidm1a2/afraidofthedark/common/spell/component/DeliveryTransitionStateBuilder.kt
package com.davidm1a2.afraidofthedark.common.spell.component

import com.davidm1a2.afraidofthedark.common.spell.Spell
import net.minecraft.entity.Entity
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.Vec3d
import net.minecraft.world.World

/**
 * Builder class used to build a DeliveryTransitionState object
 *
 * @property spell The spell that is transitioning
 * @property stageIndex The index of the spell stage that is transitioning
 * @property world The world that the transition is happening in
 * @property position The exact double vector position the transition occurred
 * @property blockPosition A rounded for of getPosition() that estimates the block the transition occurred in
 * @property direction A normalized vector pointing in the direction the transition was in
 * @property casterEntity The entity that cast the spell
 * @property entity The entity being delivered to, can be null
 * @property deliveryEntity The entity delivering the spell (ex: projectile entity), can be null
 */
class DeliveryTransitionStateBuilder {
    private lateinit var spell: Spell
    private var stageIndex = 0
    private lateinit var world: World
    private lateinit var position: Vec3d
    private lateinit var blockPosition: BlockPos
    private lateinit var direction: Vec3d
    private var casterEntity: Entity? = null
    private var entity: Entity? = null
    private var deliveryEntity: Entity? = null

    /**
     * Sets the spell to transition for
     *
     * @param spell The spell to transition for
     * @return The builder
     */
    fun withSpell(spell: Spell): DeliveryTransitionStateBuilder {
        this.spell = spell
        return this
    }

    /**
     * Sets the spell index we're transitioning from
     *
     * @param stageIndex The spell index we're transitioning from
     * @return The builder
     */
    fun withStageIndex(stageIndex: Int): DeliveryTransitionStateBuilder {
        this.stageIndex = stageIndex
        return this
    }

    /**
     * Sets the world the transition is happening in
     *
     * @param world The world the transition is happening in
     * @return The builder
     */
    fun withWorld(world: World): DeliveryTransitionStateBuilder {
        this.world = world
        return this
    }

    /**
     * Sets the position the transition is happening at
     *
     * @param position The position the transition is happening at
     * @return The builder
     */
    fun withPosition(position: Vec3d): DeliveryTransitionStateBuilder {
        this.position = position
        return this
    }

    /**
     * Sets the block position the transition is happening at
     *
     * @param blockPosition The block position the transition is happening at
     * @return The builder
     */
    fun withBlockPosition(blockPosition: BlockPos): DeliveryTransitionStateBuilder {
        this.blockPosition = blockPosition
        return this
    }

    /**
     * Sets the direction the transition is happening towards
     *
     * @param direction The direction the transition is happening towards
     * @return The builder
     */
    fun withDirection(direction: Vec3d): DeliveryTransitionStateBuilder {
        this.direction = direction
        return this
    }

    /**
     * Sets the the entity that cast the spell
     *
     * @param casterEntity The entity that cast the spell
     * @return The builder
     */
    fun withCasterEntity(casterEntity: Entity?): DeliveryTransitionStateBuilder {
        this.casterEntity = casterEntity
        return this
    }

    /**
     * Sets the delivery entity that is used to deliver the spell
     *
     * @param deliveryEntity The entity delivering the spell
     * @return The builder
     */
    fun withDeliveryEntity(deliveryEntity: Entity?): DeliveryTransitionStateBuilder {
        this.deliveryEntity = deliveryEntity
        return this
    }

    /**
     * Sets the world, position, direction, and entity based on the entity's position
     *
     * @param entity The entity to grab data from
     * @return The builder
     */
    fun withEntity(entity: Entity): DeliveryTransitionStateBuilder {
        this.entity = entity
        world = entity.entityWorld
        position = entity.getPositionEyes(1.0f)
        blockPosition = BlockPos(position)
        direction = entity.lookVec
        return this
    }

    /**
     * Sets the state of this builder based on an existing state
     *
     * @param state The state to copy from
     * @return The builder
     */
    fun copyOf(state: DeliveryTransitionState): DeliveryTransitionStateBuilder {
        spell = state.spell
        stageIndex = state.stageIndex
        world = state.world
        position = state.position
        blockPosition = state.blockPosition
        direction = state.direction
        casterEntity = state.getCasterEntity()
        entity = state.getEntity()
        deliveryEntity = state.getDeliveryEntity()
        return this
    }

    /**
     * Finishes the state object and returns it
     *
     * @return The state object
     */
    fun build(): DeliveryTransitionState {
        return DeliveryTransitionState(
            spell,
            stageIndex,
            world,
            position,
            blockPosition,
            direction,
            casterEntity,
            entity,
            deliveryEntity
        )
    }
}